### PR TITLE
[MRG] Make EEGLAB I/O more robust in copyfile_eeglab()

### LINKED ISSUE #652: copyfile_eeglab not working
### Description

I am currently attempting to copy EEGLAB files via copyfiles_eeglab with a new (BIDS compliant) name. The files are moved and renamed according to the specifications. However, when trying to open those files with MNE there is an error message:

```
Traceback (most recent call last):

  File "<ipython-input-17-8c31a528db9f>", line 1, in <module>
    raw_bids = mne.io.read_raw_eeglab('sub-001_ses-D1_task-RS2_run-02_eeg.set')

  File "C:\Users\Nils Rosjat\Anaconda3\lib\site-packages\mne\io\eeglab\eeglab.py", line 222, in read_raw_eeglab
    eog=eog, verbose=verbose, uint16_codec=uint16_codec)

  File "<decorator-gen-216>", line 24, in __init__

  File "C:\Users\Nils Rosjat\Anaconda3\lib\site-packages\mne\io\eeglab\eeglab.py", line 326, in __init__
    info, eeg_montage, update_ch_names = _get_info(eeg, eog=eog)

  File "C:\Users\Nils Rosjat\Anaconda3\lib\site-packages\mne\io\eeglab\eeglab.py", line 145, in _get_info
    ch_names, eeg_montage = _get_eeg_montage_information(eeg, has_pos)

  File "C:\Users\Nils Rosjat\Anaconda3\lib\site-packages\mne\io\eeglab\eeglab.py", line 123, in _get_eeg_montage_information
    ch_pos=dict(zip(ch_names, np.array(pos))),

TypeError: unhashable type: 'numpy.ndarray'
```

When comparing the original MATLAB structure:
before

with the newly created structure:
after

I saw that several entries were replaced by numpy array objects. This seems to happen already in

https://github.com/mne-tools/mne-bids/blob/3d96c5b2af004b57085b5c9bf884e18279707a3c/mne_bids/copyfiles.py#L531-L532

When replacing loadmat with `read_mat` (from mne.externals.pymatreader) this seems to work better, but still after saving the dict with `savemat` the same issue remains.

### Script used for conversion

**mne_bids version: 0.5**

```python
import os.path as op
import os

import mne

from mne_bids import BIDSPath
from mne_bids.copyfiles import copyfile_eeglab

parent_dir = op.abspath('..')
data_path = op.join(parent_dir,'data')

sub_dirs = ['001']
sessions = ['D1']
tasks = ['Tap']

for sub_dir in sub_dirs:
    for session in sessions:
        for task in tasks:
                sub_path = op.join(data_path, sub_dir, 'ses-' + session)
                sub_id = os.listdir(sub_path)[0].split('_')[0]
                set_name = sub_id + '_' + session + '_' + task + '_Nils_IntRef_CSD.set'
                set_file = op.join(data_path, sub_dir, 'ses-' + session,set_name)
                set_rename = 'sub-' + sub_dir + '_ses-' + session + '_task-' + task + '_desc-CSD.set'
                set_file_renamed = op.join(data_path, 'sub-' + sub_dir, 'ses-' + session, 'eeg', set_rename)
                copyfile_eeglab(set_file, set_file_renamed)
```

### PR DESCRIPTION
**Pull request by hoechenberger**
Fixes #652

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by hoechenberger** [comment]
btw pymatreader uses:
```python
            if mjv == 1:
                extra_kwargs['uint16_codec'] = uint16_codec

            raw_data = scipy.io.loadmat(fid, struct_as_record=True,
                                        squeeze_me=True, mat_dtype=False,
                                        variable_names=variable_names,
                                        **extra_kwargs)

```

https://gitlab.com/obob/pymatreader/-/blob/aef8b1452540ed778347a77cae7927db4bf81075/pymatreader/pymatreader.py#L86-92

And this is also the default for SciPy's loadmat:
https://github.com/scipy/scipy/blob/19acfed431060aafaa963f7e530c95e70cd4b85c/scipy/io/matlab/mio5.py#L177

So I have no clue why this was overridden in the first place

**Comment by hoechenberger** [comment]
I would like to add a unit test to catch future regression, but I cannot reproduce this issue without @nrosjat's data. Suggestions?

**Comment by sappelhoff** [comment]
> Suggestions

firing up EEGLAB and creating a bogus test file =) ... apparently it's running on Octave: https://sccn.ucsd.edu/wiki/Running_EEGLAB_on_Octave

or asking @nrosjat if they can provide a file

**Comment by nrosjat** [comment]
> or asking @nrosjat if they can provide a file

@hoechenberger should have received a annonymized dataset for testing by now. Just let me know in case you need something else.

**Comment by hoechenberger** [comment]
@nrosjat If it's okay with you, I'd include a small snippet (maybe a second or two or so) of the data you sent me in our set of (publicly available) testing data?

**Comment by nrosjat** [comment]
@hoechenberger you're talking about the second dataset, right? That's okay with me.

**Comment by hoechenberger** [comment]
> @hoechenberger you're talking about the second dataset, right? That's okay with me.

At first I was confused, but then I realized I had missed your second email! Perfect, yes, I'll use this dataset.

**Comment by hoechenberger** [comment]
Current PR breaks the following roundtrip test:

```python
# %%
import os.path as op

import mne
from mne.datasets import testing

import mne_bids


data_path = op.join(testing.data_path(), 'EEGLAB')
infile = op.join(data_path, 'test_raw.set')

outfile = '/tmp/foo.set'
outfile2 = '/tmp/bar.set'

mne_bids.copyfiles.copyfile_eeglab(infile, outfile)
raw = mne.io.read_raw_eeglab(outfile)
raw.save(outfile, overwrite=True)

mne_bids.copyfiles.copyfile_eeglab(outfile, outfile2)
```

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-3e69aaa69d05> in <module>
     18 raw = mne.io.read_raw_eeglab(outfile)
     19 raw.save(outfile, overwrite=True)
---> 20 mne_bids.copyfiles.copyfile_eeglab(outfile, outfile2)

~/Development/mne-bids/mne_bids/copyfiles.py in copyfile_eeglab(src, dest)
    529 
    530     # Extract matlab struct "EEG" from EEGLAB file
--> 531     mat = loadmat(src, squeeze_me=False, chars_as_strings=True,
    532                   mat_dtype=False, struct_as_record=True)
    533     if 'EEG' not in mat:

~/miniforge3/envs/mne/lib/python3.8/site-packages/scipy/io/matlab/mio.py in loadmat(file_name, mdict, appendmat, **kwargs)
    222     with _open_file_context(file_name, appendmat) as f:
    223         MR, _ = mat_reader_factory(f, **kwargs)
--> 224         matfile_dict = MR.get_variables(variable_names)
    225 
    226     if mdict is not None:

~/miniforge3/envs/mne/lib/python3.8/site-packages/scipy/io/matlab/mio4.py in get_variables(self, variable_names)
    389         mdict = {}
    390         while not self.end_of_stream():
--> 391             hdr, next_position = self.read_var_header()
    392             name = asstr(hdr.name)
    393             if variable_names is not None and name not in variable_names:

~/miniforge3/envs/mne/lib/python3.8/site-packages/scipy/io/matlab/mio4.py in read_var_header(self)
    345            position in stream of next variable
    346         '''
--> 347         hdr = self._matrix_reader.read_header()
    348         n = reduce(lambda x, y: x*y, hdr.dims, 1)  # fast product
    349         remaining_bytes = hdr.dtype.itemsize * n

~/miniforge3/envs/mne/lib/python3.8/site-packages/scipy/io/matlab/mio4.py in read_header(self)
    115         O, rest = divmod(rest, 100)  # unused, should be 0
    116         if O != 0:
--> 117             raise ValueError('O in MOPT integer should be 0, wrong format?')
    118         P, rest = divmod(rest, 10)  # data type code e.g miDOUBLE (see above)
    119         T = rest  # matrix type code e.g., mxFULL_CLASS (see above)

ValueError: O in MOPT integer should be 0, wrong format?
```

**Comment by sappelhoff** [comment]
did you try it with `chars_as_strings` False? Perhaps that's why it was set to False prior to this PR?

Perhaps changing scipy.loadmat to mne.externals.pymatreader in the copyfile code solves the issue?

**Comment by hoechenberger** [comment]
Simpler MWE to demonstrate the problem:
```python
import os.path as op
import mne
from mne.datasets import testing

import mne_bids

data_path = op.join(testing.data_path(), 'EEGLAB')
infile = op.join(data_path, 'test_raw.set')

outfile = '/tmp/foo.set'
outfile2 = '/tmp/bar.set'

raw = mne.io.read_raw_eeglab(infile)
raw.save(outfile, overwrite=True)
mne_bids.copyfiles.copyfile_eeglab(outfile, outfile2)
```

**Comment by hoechenberger** [comment]
> did you try it with `chars_as_strings` False? Perhaps that's why it was set to False prior to this PR?
> Perhaps changing scipy.loadmat to mne.externals.pymatreader in the copyfile code solves the issue?

Currently looking into both options.

**Comment by hoechenberger** [comment]
It's an MNE bug:

```python
import os.path as op
import mne
from mne.datasets import testing

import mne_bids

data_path = op.join(testing.data_path(), 'EEGLAB')
infile = op.join(data_path, 'test_raw.set')

outfile = '/tmp/foo.set'

raw = mne.io.read_raw_eeglab(infile)
raw.save(outfile, overwrite=True)
mne.io.read_epochs_eeglab(outfile)
```
reproduces the error

**Comment by hoechenberger** [comment]
Ugh I was trying to read the raw data as Epochs, stupid me… Back to square one

**Comment by hoechenberger** [comment]
No wait. even with the corrected MWE I get the same error.

```python
import os.path as op
import mne
from mne.datasets import testing

data_path = op.join(testing.data_path(), 'EEGLAB')
infile = op.join(data_path, 'test_raw.set')

outfile = '/tmp/foo.set'

raw = mne.io.read_raw_eeglab(infile)
raw.save(outfile, overwrite=True)
mne.io.read_raw_eeglab(outfile)
```

**Comment by hoechenberger** [comment]
Wait. Argh. No. We're writing a FIFF here. DAAAMMN I'm a little unfocused today!!

**Comment by hoechenberger** [comment]
> Wait. Argh. No. We're writing a FIFF here. DAAAMMN I'm a little unfocused today!!

But damn this is really unintuitive then. I mean... one might say it tripped me because I was unfocused, but this is clearly a gotcha for a beginner. If I load an EEGLAB file and call `raw.save`, I'd actually expect to get an EEGLAB file back, esp. if the output filename ends in `.set`. There should at least be a warning or something. But maybe there is and it was swallowed by my IDE.

**Comment by sappelhoff** [comment]
> But maybe there is and it was swallowed by my IDE.

I think there is one. Something along the lines "saving raw data with MNE should be done using a filename ending in raw.fif[.gz]"

File: mne_bids/copyfiles.py [hunk 1]
```diff
@@ -20,10 +20,12 @@
 
 import shutil as sh
 from datetime import datetime
+from scipy.io import savemat
```
Comment (sappelhoff) [inline mne_bids/copyfiles.py:23]: Personally I like the Python convention to have 3 import blocks: 1) builtins, 2) external, 3) local --> don't know which pep that is, but pretty sure it's part of some official style. So I would keepscipy listed near MNE (both "external")

**Comment by hoechenberger** [comment]
I'm running from one issue into the next. This is such a rabbit hole… Taking a break for now.

**Comment by hoechenberger** [comment]
Seems that `pyreadmat` is doing a little bit too much magic so that reading is great and makes it easier to use within Python, but then writing back to a .mat file using `savemat` is being impaired in some cases

**Comment by sappelhoff** [comment]
> Seems that pyreadmat is doing a little bit too much magic so that reading is great and makes it easier to use the Matlab data within Python, but then writing back to a .mat file using savemat is being impaired in some cases

mmh that sounds very possible. Makes me a bit sad though because it'd have been an easy way to also support matfiles beyond scipy's limited set of supported versions. (most importantly, v7.3).

**Comment by hoechenberger** [comment]
Ok I believe I got things to work, thanks to @cbrnr's `simplify_cells` addition to `loadmat`. However, this switch was only added in SciPy 1.5.0, which was released in June 2020. Is it reasonable to expect users who wish to use `copyfile_eeglab` to have such a recent version of SciPy installed?

**Comment by cbrnr** [comment]
Yes, SciPy 1.5.2 is in the default Anaconda channel.

**Comment by hoechenberger** [comment]
Todo: Add testing data + new test

**Review by sappelhoff — APPROVED** [review]
Looks great! Thanks a lot for going down this rabbit hole @hoechenberger :-)

**Comment by hoechenberger** [comment]
@sappelhoff But you know what's missing? A test that actually verifies that the data can be read back in properly by EEGLAB…

**Comment by hoechenberger** [comment]
> and compatibility with Matfile versions other than 4,6, and7-7.2 --> especially for 7.3, which is the EEGLAB default since some time now. So almost all new files are v7.3 😬

7.3 is HDF5, so this should be relatively simple…

Is there a compiled version of EEGLAB available for Linux? Could only find one for macOS and Windows. (Don't want to use Octave, it doesn't seem to be well-supported by EEGLAB?)

**Comment by hoechenberger** [comment]
I just checked https://github.com/sccn/eeglab to see how they do it but … there don't appear to be any tests????!

**Comment by sappelhoff** [comment]
> Is there a compiled version of EEGLAB available for Linux? Could only find one for macOS and Windows. (Don't want to use Octave, it doesn't seem to be well-supported by EEGLAB?)

there used to be one. go to: https://sccn.ucsd.edu/eeglab/downloadtoolbox.php

once you manage to get past that "registration form", you'll see: (screenshot)

image


**note** though: I tried the compiled version a couple of days ago to check whether it can be used for testing this issue --> it doesn't really work well ... there is a limited amount of GUI buttons and that's it.

**Comment by sappelhoff** [comment]
> there don't appear to be any tests????!

I guess that's MATLAB for you ... they only recently introduced a testing framework as far as I know. It may be though that SCCN has some internal testing pipeline in place that they run regularly but don't transparently document. Not sure though - pure speculation.

**Comment by hoechenberger** [comment]
> I guess that's MATLAB for you

Hm well but you could do TDD even without proper language support, no? Oh well.

**Comment by hoechenberger** [comment]
Anyway, @sappelhoff feel free to merge, this is good to go from my end.

**Comment by sappelhoff** [comment]
Thanks @hoechenberger - let's improve this one step at a time.

**Pull request merged**